src/core/polyhedral/memory_promotion_heuristic.cc

```diff
@@ -503,5 +545,114 @@ void promoteGreedilyAtDepth(
   mapCopiesToThreads(mscop, unrollCopies);
 }
 
+namespace {
+isl::val getParamValIfFixed(isl::union_set uset, int pos) {
+  auto val = isl::val::nan(uset.get_ctx());
+  for (auto set : isl::UnionAsVector<isl::union_set>(uset)) {
+    auto currentVal = set.plain_get_val_if_fixed(isl::dim_type::param, pos);
+    if (currentVal.is_nan()) {
+      return currentVal;
+    }
+    if (!val.is_nan() && val != currentVal) {
+      return isl::val::nan(uset.get_ctx());
+    }
+    val = currentVal;
+  }
+  return val;
+}
+} // namespace
+
+void promoteToRegistersBelowThreads(
+    Scop& scop,
+    const ThreadIdxxScheduleDepthState& threadIdxxScheduleDepthState,
+    size_t nRegisters) {
+  using namespace tc::polyhedral::detail;
+
+  // Assuming the mapping happens to threads happens in inverse order, i.e.
+  // the innermost loop is mapped to thread x, promote below that depth.
+
+  auto root = scop.scheduleRoot();
+
+  auto fullSched = fullSchedule(root);
+  for (const auto& kvp : threadIdxxScheduleDepthState) {
+    auto depth = kvp.second + 1;
+    auto subdomain = kvp.first;
+
+    // Collect all bands where a member is located at the given depth.
+    auto bands = bandsContainingScheduleDepth(root, depth);
+    // We may have no band members mapped to thread x in case when we
+    // force-mapped everything to one thread.
+    if (bands.size() == 0) {
+      continue;
+    }
+
+    // Keep only those bands for which the this depth was recorded.
+    std::function<bool(ScheduleTree*)> keepActive =
+        [root, subdomain](const ScheduleTree* tree) {
+          isl::union_set active = activeDomainPoints(root, tree);
+          return !active.intersect(subdomain).is_empty();
+        };
+    bands = functional::Filter(keepActive, bands);
+
+    // Make sure the band ends at thread x depth so we can promote below it.
+    bands = bandsSplitAfterDepth(bands, root, depth);
+
+    for (auto band : bands) {
+      // Find out how many threads are actually mapped.  Active domain points
+      // will involve all mapping parameters when we take them below the
+      // mapping.  Skip mapping parameters obviously mapped to 0, because they
+      // do not correspond to band members that should be fixed to obtain
+      // per-thread-group access relations.
+      auto points = activeDomainPoints(root, band);
+      size_t nMappedThreads = 0;
+      for (int j = 0; j < points.dim(isl::dim_type::param); ++j) {
+        auto id = points.get_space().get_dim_id(isl::dim_type::param, j);
+        for (size_t i = 0; i < mapping::ThreadId::kMaxDim; ++i) {
+          if (not(id == mapping::ThreadId::makeId(i))) {
```

Comment:
is this better than ```!=``` or is there an underlying compilation issue because missing operator?

Reply:
it was missing, introduced in a follow-up PR; I can cherry-pick it here